File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {

        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses, 0);

        // Build graph and calculate indegree
        for (auto& edge : prerequisites) {

            int course = edge[0];
            int prerequisite = edge[1];

            graph[prerequisite].push_back(course);
            indegree[course]++;
        }

        queue<int> que;

        // Add courses with no prerequisites
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                que.push(i);
            }
        }

        vector<int> answer(numCourses);
        int index = 0;

        // Kahn's Algorithm
        while (!que.empty()) {

            int curr = que.front();
            que.pop();

            answer[index++] = curr;

            // Process dependent courses
            for (int neighbor : graph[curr]) {

                indegree[neighbor]--;

                if (indegree[neighbor] == 0) {
                    que.push(neighbor);
                }
            }
        }

        // If not all courses were processed, a cycle exists.
        if (index != numCourses) {
            return {};
        }

        return answer;
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: list[list[int]]) -> list[int]:

        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        # Build graph and calculate indegree
        for edge in prerequisites:

            course = edge[0]
            prerequisite = edge[1]

            graph[prerequisite].append(course)
            indegree[course] += 1

        que = []

        # Add courses with no prerequisites
        for i in range(numCourses):
            if indegree[i] == 0:
                que.append(i)

        answer = [0] * numCourses
        index = 0
        front = 0

        # Kahn's Algorithm
        while front < len(que):

            curr = que[front]
            front += 1

            answer[index] = curr
            index += 1

            # Process dependent courses
            for neighbor in graph[curr]:

                indegree[neighbor] -= 1

                if indegree[neighbor] == 0:
                    que.append(neighbor)

        # If not all courses were processed, a cycle exists.
        if index != numCourses:
            return []

        return answer